### ISSUE TITLE
SPIKE: Pendo Visitor and Account

### ISSUE THREAD

**Issue by JamWils** [weaveworks/weave-gitops#2485]
We want to use a product tracking tool to help us better understand how the product is being used on a day-to-day basis.  This spike is intended to outline how we will group users and organizations, in Pendo these are respectively visitors and accounts.

**For the CLI:**

We generate a random hash for each user to help us understand growth in adoption and measure the success of the project.  We would want this hash to be consistent on each run when we send a track event. We have neither the interest or mechanism by which we could identify a person based on our collected data.   

If we go this route would you also be able to set some orgId (i.e. accountId) as well.  <-- This needs to be validated to see if it is possible.

** For the Dashboard:**

We want to be able to support both the default user account (if 10 people are using this so be it that we cannot tell the difference) as well as OIDC users.  Outline what the path is for visitorId and accountId (I think we discussed cluster-id for the account).  How can we hash these values so:

1. We do not know what the original value is?
2. The same value is always hashed to the same value for consistent tracking.  

E.g. we don't need to know a company has named their cluster `aws-secret-cluster`, it's enough to have the tag `34298734aabccddd` as the cluster name.

Outcome of this spike: clearly defining how we are setting `visitorId` and `accountId`.

**Comment by ozamosi** [comment]
**CLI:**
I cannot find support in Pendo for CLIs, and I cannot find support for golang. There's support for iOS, Android and web, and there's support for interfacing with those platforms directly or through flutter/xamarin/react native.

I can't find any way Pendo supports custom integrations - am I missing something?

We _can_ use what's roughly the go-checkpoint API (which can send an event every time we run the CLI), and we can push that onwards through :waves-hand-dismissively: data pipeline integrations to :rolls-eyes: useful places.

As for a **visitorID** - https://github.com/denisbrodbeck/machineid seems to implement a cross-platform way to get a unique value from a machine that doesn't change until the computer is reinstalled. The raw id is sensitive, but if it's hashed enough it would save us from having to manage our own state - that repo suggests wrapping the value in hmac-sha256.

As for **accountID**. We could connect to the current context cluster and look for the UID of an object we "think we have access to" (the gitops dashboard, perhaps?), but it feels both a bit creepy and fragile.

We could grab the default cluster name and hash it - it'd both risk 2 users having different names for the same cluster (e.g. I can call a cluster "prod" and you can call it "production" - should be rare with most modern tooling writing a whole context at once, though), and it'd merge 2 users having the same name for different clusters (e.g. "the kind-kind problem"). Depending on what we want to do with this data, maybe this isn't a problem?

**Dashboard**

**VisitorID**: Grab `UserPrincipal.ID` from the auth object. This _must be hashed_ because it's probably an email in the OIDC case.

** AccountID**: a cluster doesn't really have a name - your kubeconfig lets you reference it by a name, but the name isn't available inside. I suggest using the UID of some object we have access to. We have access to read all namespaces and 2 secrets - if we grabbed the UID of `kube-system`, we can be reasonably sure it won't be recreated unless the cluster is. It's random and doesn't really mean much, we should still hash it.

**Hashing**

We should not just call sha256() on the string and call it a day. We need an application-unique constant to make sure our hashes doesn't match with anybody else's - e.g. if the data were to leak, a sha256() on just your email can be cross-referenced with other leaked data, whereas e.g. hmac or sha3 on email + random constant cannot. The "random constant" has to be part of our application and has to be the same for as long as we want to keep the data consistent - rotating the key would cause all ids to change.


**tl;dr**

CLI visitorID: use https://github.com/denisbrodbeck/machineid 
CLI accountID: `kubeClient.ClusterName`
Dashboard visitorID: pkg/server/auth.UserPrincipal.ID
Dashboard accountID: `cs.clientsFactory.GetClustersNamespaces()["Default"]`, find `kube-system`, call `GetUID()`

Before sharing, send all IDs through the mac example here: https://pkg.go.dev/golang.org/x/crypto/sha3#example-package-Mac

**Comment by JamWils** [comment]
There is API support with Track Events so we don't need an SDK we can just make a POST request.  An action item for me is to see if we have access to the API via Pendo Free.


https://support.pendo.io/hc/en-us/articles/360032294151-Track-Events

https://developers.pendo.io/docs/?bash#track-events

**Comment by ozamosi** [comment]
Start doing the work in #2451
